### DISCUSSION TITLE
Identifying connected clusters

### DISCUSSION THREAD

**Discussion by vishankkumar** [kip-hart/MicroStructPy#42] [Ideas]
Is it possible to create a flag that identify the connected spheres, where connection is defined if the distance between two particles of radius r1 and r2 is less than r1 + r2? I wanted to generate a 2 phase microstructure along with additional phases for connected clusters of phase 1 and phase 2.

**Comment by kip-hart** [comment]
Hi Vishan,

It's possible to do this, although it may take the code a long time to run if it has to check if sphere "i" and sphere "j" are close enough to add a phase between them. Another approach would be to create a PolyMesh from the spheres of phase 1 and 2, then add spheres of phase 3+ to the facets between phases 1 and 2. Here's the steps:

1. Create a SeedList of phases 1 and 2 (`seeds_12`)
2. Position those seeds in the domain
3. Create a PolyMesh (`pmesh_12`) from `seeds_12`
4. For each facet neighbor pair (`pmesh_12.facet_neighbors`) in the PolyMesh:
    a. Check if phase numbers (`pmesh_12.phase_numbers`) are unique
    b. Check if the distance between the seeds is close enough, etc
    c. Create a seed (or multiple seeds) of phase 3+ if (a) and (b) checks pass
    d. Position the seed(s) on that facet
    e. Append seed to a new SeedList (`seeds_3`)
5. Define new SeedList `seeds_all = seeds_12 + seeds_3`
6. Create a PolyMesh (`pmesh_all`) from `seeds_all`
7.  Create a TriMesh from `pmesh_all`

I wrote something similar to this in the [`foam.py`](https://docs.microstructpy.org/en/v1.4.10/examples/package/foam.html) example. First a PolyMesh is created from the voids, then circles of the foam phase are added to the facets between the voids.

![foam example](https://docs.microstructpy.org/en/latest/_images/trimesh23.png)

Let me know how it goes and feel free to follow up. I'm happy to help.

Best,
Kip
